src/components/verification-grid/verification-grid.ts

```diff
@@ -314,22 +347,116 @@ export class VerificationGridComponent extends WithShoelace(AbstractComponent(Li
+   *    2. All tasks on the selected tile is completed
+   *
+   * You can enter this mode at any time, by simply completing all tasks on a
+   * single tile.
+   * Once in this mode, there is some special functionality like the first tile
+   * of each new page being automatically selected.
+   */
+  private singleDecisionMode = false;
+
   private requiredClassificationTags: Tag[] = [];
   private requiredDecisions: RequiredDecision[] = [];
   private hiddenTiles = 0;
   private showingSelectionShortcuts = false;
-  private selectionHead: number | null = null;
   private anyOverlap = signal<boolean>(false);
   private gridController?: DynamicGridSizeController<HTMLDivElement>;
   private paginationFetcher?: GridPageFetcher;
+
+  private highlightSelectionAnimation = newAnimationIdentifier("highlight-selection");
   private highlight: HighlightSelection = {
     start: { x: 0, y: 0 },
     current: { x: 0, y: 0 },
     highlighting: false,
     observedElements: [],
   };
 
-  private highlightSelectionAnimation = newAnimationIdentifier("highlight-selection");
+  private focusHead: number | null = null;
+  private _rangeSelectionHead: number | null = null;
+
+  /**
+   * Where range selection will start from.
+   * This pointer moves independently from the focus head.
+   */
+  private get rangeSelectionHead() {
+    return this._rangeSelectionHead;
+  }
+
+  private set rangeSelectionHead(value: number | null) {
+    // Note that the bounds check upper limit is the populated tile count
+    // subtract one.
+    // This is because the populatedTileCount is indexed from 1 while the range
+    // selection value is indexed from 0.
+    const upperBound = this.populatedTileCount - 1;
+    const isInBounds = value === null || (value >= 0 && value <= upperBound);
+    if (!isInBounds) {
+      console.error(`new range selection value: '${value}' is not valid. Value must be in range [0,${upperBound}]`);
+      return;
+    }
+
+    this._rangeSelectionHead = value;
+
+    // When unsetting the selection head (e.g. with ESC key) we want to keep the
+    // focus head at the same index, so that if you start moving again, it will
+    // start from where the current focus head is (where you deselected from).
+    if (value !== null) {
+      this.focusHead = value;
+    }
+  }
+
+  private get focusLeftIndex(): number {
+    return this.focusHead === null ? 0 : Math.max(this.focusHead - 1, 0);
+  }
+
+  private get focusRightIndex(): number {
+    return this.focusHead === null ? 0 : Math.min(this.focusHead + 1, this.lastTileIndex);
+  }
+
+  private get focusUpIndex(): number {
+    if (this.focusHead === null) {
+      return 0;
+    }
+
+    // If the selection head is on the top row, pressing up should have no
+    // action.
+    const proposedIndex = this.focusHead - this.columns;
+    if (proposedIndex < 0) {
+      return this.focusHead;
+    }
+
+    return proposedIndex;
+  }
+
+  private get focusDownIndex(): number {
```

Comment:
these are great abstractions but _focus_ is a verb, and it makes it sound like the methods are doing something.

suggestion `next[Left|Right|Up|Down]Index`